src/platform/Darwin/BleConnectionDelegateImpl.mm

```diff
@@ -280,21 +276,27 @@ - (BOOL)isConnecting
     return _currentMode == kConnecting;
 }
 
-- (void)resetTimer
+- (void)setupTimer:(uint64_t)timeout
 {
-    if ([self isConnecting]) {
-        auto timeout = static_cast<int64_t>(kScanningWithDiscriminatorTimeoutInSeconds * NSEC_PER_SEC);
-        dispatch_source_set_timer(_timer, dispatch_walltime(nullptr, timeout), DISPATCH_TIME_FOREVER, 5 * NSEC_PER_SEC);
-    } else if ([self isScanningWithoutDelegate]) {
-        auto timeout = static_cast<int64_t>(kScanningWithoutDelegateTimeoutInSeconds * NSEC_PER_SEC);
-        dispatch_source_set_timer(_timer, dispatch_walltime(nullptr, timeout), DISPATCH_TIME_FOREVER, 5 * NSEC_PER_SEC);
-    } else if ([self isScanning]) {
-        dispatch_source_cancel(_timer);
-    } else {
-        // It should not happens.
-        [self stop];
-        [self dispatchConnectionError:CHIP_ERROR_INCORRECT_STATE];
+    if (_timer) {
+        [self clearTimer];
     }
+
+    _timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, _workQueue);
+    dispatch_source_set_event_handler(_timer, ^{
+        [self stop];
+        [self dispatchConnectionError:BLE_ERROR_APP_CLOSED_CONNECTION];
+    });
+
+    auto value = static_cast<int64_t>(timeout * NSEC_PER_SEC);
+    dispatch_source_set_timer(_timer, dispatch_walltime(nullptr, value), DISPATCH_TIME_FOREVER, 5 * NSEC_PER_SEC);
+    dispatch_resume(_timer);
+}
+
+- (void)clearTimer
```

Comment:
It might be worth pushing the `if (_timer)` check into this function instead of doing it at every callsite.

Reply:
Done.